src/lib/CheapReduction.hs

```diff
@@ -193,6 +193,12 @@ instance CheaplyReducibleE Atom Atom where
         Failure _ -> do
           reportSynthesisFail ty
           return $ Con $ DictHole ctx ty
+    TabPi (TabPiType (b:>IxType ixTy dict) resultTy) -> do
+      ixTy' <- cheapReduceE ixTy
+      dict' <- cheapReduceE dict
+      withFreshBinder (getNameHint b) (IxType ixTy' dict') \b' -> do
+        resultTy' <- extendSubst (b@>Rename (binderName b')) $ cheapReduceE resultTy
+        return $ TabPi $ TabPiType (b':>IxType ixTy' dict') resultTy'
```

Comment:
Is this enough? What if the `TabPi` is nested inside a `Pi` type? The `Pi` would go through the default path of traversing and simplifying the names in an expr, leaving the dict unmodified. What is the invariant you're going for after cheap reduction?

Reply:
Yes, I think you're right. We can't just fall back to reducing the names like we used to (it's possible it was broken already, because it would miss phantom dictionaries inside Pi types). I'll fix cheap reduction next. I'll clarify the dictionary synthesis invariant and handle all the cases properly.

Reply:
Seems like this still hasn't been fixed? I'm not sure why do we even need the special rule for table pis here?

Reply:
Yes, this slipped through. I had a large patch which addressed it but I eventually abandoned it: https://github.com/google-research/dex-lang/tree/corporeal-ix-abandoned-path

Reply:
btw, in case it's not obvious, this is also broken on main. The problem is that phantom dictionaries might not be discovered when we just traverse the names.

Reply:
For example, if there are dictionary holes in a lambda term, then this case ignores them:

```
    -- Don't try to eagerly reduce lambda bodies. We might get stuck long before
    -- we have a chance to apply tham. Also, recursive traversal of those bodies
    -- means that we will follow the full call chain, so it's really expensive!
    Lam _   -> substM a
```